### PR TITLE
fix: Force UTF-8 encoding in C# lit runner

### PR DESCRIPTION
**Pull request by cpitclaudel**
This is intended to make it possible to test Unicode output.  I skipped Java because Java needs a separate patch (since it doesn't respect the current encoding / codepage, see https://github.com/dafny-lang/dafny/issues/2976)

### CODE REVIEW

File: Source/XUnitExtensions/Lit/ShellLitCommand.cs [hunk 1]
```diff
@@ -33,13 +33,70 @@ public class ShellLitCommand : ILitCommand {
...
+      // With Acute) should output a single byte, #x82.  But if the codepage is
+      // 1252 (Latin 1), then that same program should output the single
+      // byte `#xE9`.  Finally, if the codepage is 65001 (UTF8), then that same
+      // program should output `#xC3 #xA9`.
+      //
+      // Thankfully, the C# runtime takes care of this for us.  It uses
+      // `GetConsoleCP` to determine what encoding to use, then uses that when
+      // `Console.WriteLine` is called (and replaces unsupported characters by
+      // the fallback character `?`).  Here is a concrete demo, with character
+      // U+20AC (€) (which is supported by codepage 85001 but not 850):
```
Comment (robin-aws) [inline Source/XUnitExtensions/Lit/ShellLitCommand.cs:62]:
```suggestion
      // U+20AC (€) (which is supported by codepage 65001 but not 850):
```

File: Source/XUnitExtensions/Lit/ShellLitCommand.cs [hunk 2]
```diff
@@ -33,13 +33,70 @@ public class ShellLitCommand : ILitCommand {
...
+      // output, as is done below.
+      //
+      // … but this is not enough, because dotnet run (or XUnit) also resets the
+      // codepage.  That's a bug there, and in the meantime the best hack we
+      // have is to override that codepage by setting `OutputEncoding`.
+
+      process.StartInfo.CreateNoWindow = false;
+      Console.OutputEncoding = System.Text.Encoding.UTF8;
+
+      // Finally, for Java + Ubuntu, we make sure to forward LANG:
```
Comment (robin-aws) [inline Source/XUnitExtensions/Lit/ShellLitCommand.cs:95]:
```suggestion
      // Finally, for Java + Ubuntu, we make sure to set LANG:
```

File: Test/metatests/OutputEncoding.dfy [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// RUN: %baredafny verify %args "%s" > "%t"
+// RUN: %baredafny run --no-verify --target=cs %args "%s" >> "%t"
+// RUN: %baredafny run --no-verify --target=js %args  "%s" >> "%t"
+// RUN: %baredafny run --no-verify --target=go %args  "%s" >> "%t"
+// RUN: %baredafny run --no-verify --target=py %args  "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
```
Comment (robin-aws) [inline Test/metatests/OutputEncoding.dfy:6]: Missing the `--target=java` case.

I'm just fine with ignoring C++ for now, but could we use `%testDafnyForEachCompiler` and just add an `int` somewhere (so C++ can cleanly opt out of this test case)?

**Review by robin-aws — CHANGES_REQUESTED** [review]
Very nice work getting this working (again) :)

File: Test/metatests/OutputEncoding.dfy (hunk 3 above)
Comment (cpitclaudel) [reply Test/metatests/OutputEncoding.dfy:6]: This is on purpose, because Java doesn't work.
The next patch will make it work (by forcing UTF-8)

Comment (robin-aws) [reply Test/metatests/OutputEncoding.dfy:6]: Never mind, reading comprehension fail on the PR description. :)

**Pull request merged**